pkg/cluster/cert.go

```diff
@@ -231,6 +193,33 @@ func recreateSecret(ctx context.Context, c client.Client, existingSecret, newSec
 	return nil
 }
 
+// generateCertSecret creates a secret if it does not exist;  recreate this secret if type not match; update data if outdated.
+func generateCertSecret(ctx context.Context, c client.Client, certSecret *corev1.Secret, secretName, namespace string) error {
+	existingSecret := &corev1.Secret{}
+	err := c.Get(ctx, client.ObjectKey{Name: secretName, Namespace: namespace}, existingSecret)
+	if err != nil {
+		if k8serr.IsNotFound(err) {
+			// Secret does not exist, create it
+			if createErr := c.Create(ctx, certSecret); createErr != nil {
+				return fmt.Errorf("failed creating certificate secret: %w", createErr)
+			}
+		} else {
+			return fmt.Errorf("failed getting certificate secret: %w", err)
+		}
+	} else if existingSecret.Type != certSecret.Type {
+		// Secret exists but with a different type, delete and create it again
+		if recreateSecret(ctx, c, existingSecret, certSecret) != nil {
+			return errors.New("failed to recreate secret with type corrected")
+		}
+	}
+	if isSecretOutdated(existingSecret.Data, certSecret.Data) {
```

Comment:
If I properly understood the code, the `isSecretOutdated` function is invoked even if the secret is created or recreated but since the `existingSecret` is never updated (and there is no reason to do it), then it could lead to a spurious `Update` so the two if/else should probably return early.

Reply:
you are right. 
- in the successful Create() case, it should return immediately
- in the recreateSecret() case, it does not need to check isSecretOutdated() as well